File: puzzles/medium/kotlin/src/bender1.kt
import java.util.Scanner

const val UNBREAKABLE_WALL = '#'
const val BREAKABLE_WALL = 'X'
const val START = '@'
const val FINISH = '$'
const val SOUTH = 'S'
const val EAST = 'E'
const val NORTH = 'N'
const val WEST = 'W'
const val BEER = 'B'
const val INVERTER = 'I'
const val TELEPORTER = 'T'
const val EMPTY = ' '

fun main(args : Array<String>) {
  val city = City()
  val bender = Bender()
  readGameInput(city, bender)
  city.printGrid()
  gameLoop(city, bender)
}

fun readGameInput(city: City, bender: Bender) {
  val input = Scanner(System.`in`)
  city.height = input.nextInt()
  city.width = input.nextInt()
  if (input.hasNextLine()) {
    input.nextLine()
  }
  for (y in 0 until city.height) {
    val row = input.nextLine()
    val cellList = ArrayList<Cell>()
    for (x in 0 until row.length) {
      val symbol = row[x]
      val cell = Cell(symbol)
      cellList.add(cell)
      if (symbol == START) {
        bender.point.x = x
        bender.point.y = y
      } else if (symbol == TELEPORTER) {
        if (!city.teleporter) {
          city.teleporter1 = Point(x, y)
          city.teleporter = true
        } else {
          city.teleporter2 = Point(x, y)
        }
      }
    }
    city.grid.add(cellList)
  }
}

fun resetGridState(city: City) {
  for (y in 0 until city.height) {
    for (x in 0 until city.width) {
      city.grid[y][x].reset()
    }
  }
}

fun currentSymbol(city: City, bender: Bender): Char {
  return city.grid[bender.point.y][bender.point.x].symbol
}

fun isValidMove(city: City, bender: Bender, direction: Direction): Boolean {
  return when (direction) {
    Direction.SOUTH -> isValidPosition(city, bender, bender.point.x, bender.point.y + 1)
    Direction.EAST -> isValidPosition(city, bender, bender.point.x + 1, bender.point.y)
    Direction.NORTH -> isValidPosition(city, bender, bender.point.x, bender.point.y - 1)
    else -> isValidPosition(city, bender, bender.point.x - 1, bender.point.y) // WEST
  }
}

fun isValidPosition(city: City, bender: Bender, x: Int, y: Int): Boolean {
  return x >= 0 && x < city.width &&
      y >= 0 && y < city.height &&
      city.grid[y][x].symbol != UNBREAKABLE_WALL && (bender.breakMode || city.grid[y][x].symbol != BREAKABLE_WALL)
}

fun updateBenderState(city: City, bender: Bender)  {
  val symbol = currentSymbol(city, bender)
  if (symbol == START || symbol == SOUTH) {
    bender.direction = Direction.SOUTH
  } else if (symbol == EAST) {
    bender.direction = Direction.EAST
  } else if (symbol == NORTH) {
    bender.direction = Direction.NORTH
  } else if (symbol == WEST) {
    bender.direction = Direction.WEST
  } else if (symbol == BEER) {
    bender.breakMode = !bender.breakMode
  } else if (symbol == INVERTER) {
    bender.reverse = !bender.reverse
    bender.priorites.reverse()
  } else if (symbol == TELEPORTER) {
    if (bender.point == city.teleporter1) {
      bender.point = city.teleporter2.copy()
    } else {
      bender.point = city.teleporter1.copy()
    }
  }
}

fun addMove(city: City, bender: Bender) {
  updateBenderState(city, bender)

  if (isValidMove(city, bender, bender.direction)) {
    bender.addMove()
  } else {
    for (direction in bender.priorites) {
      if (isValidMove(city, bender, direction)) {
        bender.direction = direction
        bender.addMove()
        break
      }
    }
  }

  val cell = city.grid[bender.point.y][bender.point.x]

  if (cell.visited && Triple(cell.reverse, cell.breakMode, cell.direction) == Triple(bender.reverse, bender.breakMode, bender.direction)) {
    city.loop = true
  } else {
    cell.visited = true
    cell.reverse = bender.reverse
    cell.breakMode = bender.breakMode
    cell.direction = bender.direction
  }
}

fun gameLoop(city: City, bender: Bender) {
  while (currentSymbol(city, bender) != FINISH) {
    addMove(city, bender)

    if (city.loop) {
      break
    }

    if (bender.breakMode && currentSymbol(city, bender) == BREAKABLE_WALL) {
      city.grid[bender.point.y][bender.point.x] = Cell(EMPTY)
      resetGridState(city)
    }
  }

  if (city.loop) {
    println("LOOP")
  } else {
    for (move in bender.moves) {
      println(move)
    }
  }
}

data class Point(var x: Int, var y: Int)

enum class Direction {
  SOUTH,
  EAST,
  NORTH,
  WEST
}

class Cell(val symbol: Char) {
  var visited = false
  var direction = Direction.SOUTH
  var reverse = false
  var breakMode = false

  fun reset() {
    visited = false
    direction = Direction.SOUTH
    reverse = false
    breakMode = false
  }

  override fun toString(): String {
    return symbol.toString()
  }
}

class Bender {
  val priorites = arrayListOf(Direction.SOUTH, Direction.EAST, Direction.NORTH, Direction.WEST)
  var direction = Direction.SOUTH
  var reverse = false
  var breakMode = false
  var point = Point(0, 0)
  val moves = ArrayList<Direction>()

  fun addMove() {
    when (direction) {
      Direction.SOUTH -> point.y++
      Direction.EAST -> point.x++
      Direction.NORTH -> point.y--
      else -> point.x-- // WEST
    }
    moves.add(direction)
  }
}

class City {
  val grid = ArrayList<ArrayList<Cell>>()
  var teleporter = false
  var teleporter1 = Point(0, 0)
  var teleporter2 = Point(0, 0)
  var height = 0
  var width = 0
  var loop = false

  fun printGrid() {
    System.err.println("$height $width")

    for (y in 0 until height) {
      for (x in 0 until width) {
        System.err.print(grid[y][x].toString())
      }
      System.err.println()
    }
  }
}

File: selected-puzzles/medium/bender1/bender1.py
STRONG_WALL = "#"
WEAK_WALL = "X"
START = "@"
FINISH = "$"
SOUTH = "S"
EAST = "E"
NORTH = "N"
WEST = "W"
BEER = "B"
INVERTER = "I"
TELEPORTER = "T"
EMPTY = " "


class Cell:
    def __init__(self, symbol):
        self.symbol = symbol
        self.reset()

    def reset(self):
        self.reverse = None
        self.break_mode = None
        self.direction = None


class Bender:
    def __init__(self):
        self.priorities = ["SOUTH", "EAST", "NORTH", "WEST"]
        self.direction = "SOUTH"
        self.reverse = False
        self.break_mode = False
        self.x = 0
        self.y = 0
        self.moves = []

    def add_move(self):
        if self.direction == "SOUTH":
            self.y += 1
        elif self.direction == "EAST":
            self.x += 1
        elif self.direction == "NORTH":
            self.y -= 1
        else:  # WEST
            self.x -= 1
        self.moves.append(self.direction)


class City:
    def __init__(self):
        self.teleporter1 = None
        self.teleporter2 = None
        self.height = 0
        self.width = 0
        self.loop = False


grid = []
bender = Bender()
city = City()


def read_game_input():
    city.height, city.width = [int(i) for i in input().split()]

    for y in range(city.height):
        row = input()
        cell_list = []
        for x, symbol in enumerate(row):
            cell = Cell(symbol)
            cell_list.append(cell)
            if symbol == START:
                bender.x = x
                bender.y = y
            elif symbol == TELEPORTER:
                if city.teleporter1 is None:
                    city.teleporter1 = (x, y)
                else:
                    city.teleporter2 = (x, y)
        grid.append(cell_list)


def reset_grid_state():
    for y in range(city.height):
        for x in range(city.width):
            grid[y][x].reset()


def current_symbol():
    return grid[bender.y][bender.x].symbol


def is_valid_move(direction):
    if direction == "SOUTH":
        return is_valid_position(bender.x, bender.y + 1)
    elif direction == "EAST":
        return is_valid_position(bender.x + 1, bender.y)
    elif direction == "NORTH":
        return is_valid_position(bender.x, bender.y - 1)
    else:  # WEST
        return is_valid_position(bender.x - 1, bender.y)


def is_valid_position(x, y):
    s = grid[y][x].symbol
    x_valid = x >= 0 and x < city.width
    y_valid = y >= 0 and y < city.height
    s_valid = s != STRONG_WALL and (s != WEAK_WALL or bender.break_mode)
    return x_valid and y_valid and s_valid


def update_bender_state():
    if current_symbol() in (START, SOUTH):
        bender.direction = "SOUTH"
    elif current_symbol() == EAST:
        bender.direction = "EAST"
    elif current_symbol() == NORTH:
        bender.direction = "NORTH"
    elif current_symbol() == WEST:
        bender.direction = "WEST"
    elif current_symbol() == BEER:
        bender.break_mode = not bender.break_mode
    elif current_symbol() == INVERTER:
        bender.reverse = not bender.reverse
        bender.priorities.reverse()
    elif current_symbol() == TELEPORTER:
        if (bender.x, bender.y) == city.teleporter1:
            bender.x, bender.y = city.teleporter2
        else:
            bender.x, bender.y = city.teleporter1


def add_move():
    update_bender_state()

    if is_valid_move(bender.direction):
        bender.add_move()
    else:
        for direction in bender.priorities:
            if is_valid_move(direction):
                bender.direction = direction
                bender.add_move()
                break

    cell = grid[bender.y][bender.x]
    cell_state = (cell.reverse, cell.break_mode, cell.direction)
    bender_state = (bender.reverse, bender.break_mode, bender.direction)
    if cell_state == bender_state:
        city.loop = True
    else:
        cell.reverse = bender.reverse
        cell.break_mode = bender.break_mode
        cell.direction = bender.direction


def game_loop():
    while current_symbol() != FINISH:
        add_move()

        if city.loop:
            break

        if bender.break_mode and current_symbol() == WEAK_WALL:
            grid[bender.y][bender.x] = Cell(EMPTY)
            reset_grid_state()

    if city.loop:
        print("LOOP")
    else:
        for move in bender.moves:
            print(move)


read_game_input()
game_loop()
